### PR TITLE
Link prediction

### PR DESCRIPTION
**Pull request by will-gerard**

### CODE REVIEW

File: scripts/link_prediction.py [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+	first = True #flag to keep track of whether or not we are looking at the first line of the file
+	num_nodes = 0 #number of nodes in the graph
+	num_dimensions = 0 #number of dimensions in the embedded space
+	node_embeddings_matrix = None #the embedding similarity matrix, not initialized until we read the dimensions
+	with open(path_to_embeddings) as fp:
+		#read the number of nodes and dimension from the first line of the file
+		num_nodes, num_dimensions = [int(x) for x in next(fp).split()]
+
+		#we need to store the node embeddings in order in the matrix
+		size = (10830, num_dimensions)
```
Comment (mzwong) [inline scripts/link_prediction.py]: Probably shouldn't hard code # of nodes (10830). If we need to hard code a variable, at least make it a variable at the top of the script

File: scripts/link_prediction.py [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+		#node_embeddings_matrix = np.empty((size))
+		node_embeddings_matrix = lil_matrix(size)
+
+		#update the rows in the embeddings matrix with the appropriate values
+		for line in fp:
+			nums = [float(x) for x in line.split()]
+			node_index = int(nums[0]) #- 1 if there is no node 0 (like in the karate graph)
+			#This check is necessary become currently during the removal of edges some nodes are completely lost,
+			#so the number of nodes embedded is lower than the total number of nodes in the original graph
+			if node_index < num_nodes:
```
Comment (mzwong) [inline scripts/link_prediction.py]: remove this check, no longer necessary

File: scripts/link_prediction.py [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+				embedded_vector = nums[1:]
+				node_embeddings_matrix[node_index] = embedded_vector
+				#print("row")
+
+	#now time to initialize similarity matrix S to a matrix of the appropriate size, filled with zeros
+	size = (10830, 10830)
```
Comment (mzwong) [inline scripts/link_prediction.py]: don't hardcode, use variable at the top

File: scripts/link_prediction.py [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+	S = np.zeros(size)
+
+	for i in range(0, 10830):
```
Comment (mzwong) [inline scripts/link_prediction.py]: same here

File: scripts/separate_edge_lists.py [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+			e = (first, second)
+			full_edge_set.add(e)
+
+	#get the number of elements in the testing set:
+	num_edges = len(full_edge_set)
+	L_train = math.ceil(training_percentage * num_edges)
+	L_test = num_edges - L_train
+
+	#TODO: make this number a parameter
+	num_nodes = 34
```
Comment (mzwong) [inline scripts/separate_edge_lists.py]: Should this be a variable at the top?

File: scripts/link_prediction.py [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+	@return y_scores 	A list containing the scores for the removed edges followed by the scores for the nonexistent edges
+	'''
+
+	y_scores = []
+	for i in range(len(E_test)):
+		e = E_test[i]
+		node1 = e[0]
+		node2 = e[1]
+		#need to subtract one if there is no node 0
+		score = S[node1][node2]
```
Comment (mzwong) [inline scripts/link_prediction.py]: all these subtract ones can be fixed if you subtract a variable from them. Set the variable up top like so:
```
ZERO_INDEXING_NODES = True
if ZERO_INDEXING_NODES:
  offset = 1
else:
  offset = 0
....
score = S[node1][node2] - offset

File: scripts/link_prediction.py [hunk 7]
```diff
@@ -0,0 +1,131 @@
+from sklearn.metrics import roc_auc_score
+from sklearn.metrics.pairwise import cosine_similarity
+from scipy.sparse import lil_matrix
+
+import numpy as np
+
+def read_edge_array(path_to_edgelist):
+	'''
```
Comment (mzwong) [inline scripts/link_prediction.py]: <nit> why are your tabs so big lmao? 8 spaces seems excessive but yolo go big or go home i guess

File: scripts/separate_edge_lists.py [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+			E_fake.append(edge1)
+			count+=1
+
+	#select L random edges to remove
+	#get the full list of the edges, we will remove some to create the testing list
+	full_edge_list = list(full_edge_set)
+	E_train = []
+	E_test = []
+	#randomly select the edges to remove
+	#TODO: check to make sure removing these edges leaves the graph connected
```
Comment (mzwong) [inline scripts/separate_edge_lists.py:68]: we need to decide how necessary it is to keep the graph connected

**Review by mzwong — CHANGES_REQUESTED** [review]

File: scripts/link_prediction.py (hunk 7 above)
Comment (will-gerard) [reply scripts/link_prediction.py]: Are those 8 spaces long on your computer? On mine they are 4 spaces.

File: scripts/link_prediction.py (hunk 1 above)
Comment (will-gerard) [reply scripts/link_prediction.py]: Done

File: scripts/link_prediction.py (hunk 2 above)
Comment (will-gerard) [reply scripts/link_prediction.py]: Done

File: scripts/link_prediction.py (hunk 3 above)
Comment (will-gerard) [reply scripts/link_prediction.py]: Done

File: scripts/link_prediction.py (hunk 4 above)
Comment (will-gerard) [reply scripts/link_prediction.py]: Done

File: scripts/link_prediction.py (hunk 6 above)
Comment (will-gerard) [reply scripts/link_prediction.py]: Done

File: scripts/separate_edge_lists.py (hunk 5 above)
Comment (will-gerard) [reply scripts/separate_edge_lists.py]: Changed to a mandatory parameter the function accepts

**Review by mzwong — APPROVED** [review]

**Pull request merged**